### PR TITLE
Fix S1607: Do not raise when reason is specified in the Ignore attrib…

### LINKED ISSUE #1124: Fix S1607: Should be applied for test classes as well
### Description
Since tests can be ignored by putting the [Ignore] attribute on the test class, I think it would make sense that the rule as also checking those cases as well.
Now it's just checking if the [Ignore] is put on the test method level.

[RSPEC-1607](https://jira.sonarsource.com/browse/RSPEC-1607)

### Repro steps

Put [Ignore] next to the [TestClass] attribute and all the tests are ignored but the rule is not triggered.

### Expected behavior
Rule should be triggered for the test class as well.

### Additional improvement to be considered
- Another rule which would fail if the Ignore attribute is put on the class level would also be satisfactory.

- In our team people are putting the comment for ignoring the test as a parameter of the Ignore attribute. Maybe the rule could also take that comment as a valid fix.
For example:
[Ignore("Test is ignored because....")]

### LINKED ISSUE #1126: Fix S1607: Should not raise issues for [Ignore] attribute with a parameter
The latest version of MSTest adds a string parameter to the Ignore attribute to specify a reason. S1607 should not raise issues when the parameter is specified.

Requested in #1124

### PR DESCRIPTION
**Pull request by valhristov**
…ute, extend support for test classes

Fix #1124 
Fix #1126

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs [hunk 1]
```diff
@@ -13,18 +13,50 @@ public void Foo1()
         }
 
         [TestMethod]
-        [Ignore] // Reenable when something
+        [Ignore] // Ignored because reasons...
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: 🤣

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs [hunk 2]
```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
...
+                var testMethodOrClass = c.SemanticModel.GetDeclaredSymbol(attributeTarget);
+
+                if (testMethodOrClass != null &&
+                    testMethodOrClass.GetAttributes().Any(IsTestOrTestClassAttribute) &&
+                    !testMethodOrClass.GetAttributes().Any(IsWorkItemAttribute))
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, attribute.GetLocation()));
+                }
+            },
+            SyntaxKind.Attribute);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: Converted to AttributeSyntax action to support test class and test method with the same code

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs [hunk 3]
```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    if (!c.IsTest())
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: Should I include this check?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs [hunk 4]
```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
...
+            },
+            SyntaxKind.Attribute);
+        }
+
+        private bool HasReasonPhrase(AttributeSyntax ignoreAttributeSyntax) =>
+            ignoreAttributeSyntax.ArgumentList?.Arguments.Count > 0; // Any ctor argument counts are reason phrase
+
+        private static bool HasTrailingComment(SyntaxNode ignoreAttributeSyntax) =>
+            ignoreAttributeSyntax.Parent.GetTrailingTrivia()
+                .Any(trivia => trivia.IsKind(SyntaxKind.SingleLineCommentTrivia));
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: I know this was like this but maybe we should also handle multiline comments `/* Ignored because something */`

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs [hunk 5]
```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
...
+                var attributeTarget = attribute.Parent?.Parent;
+                if (attributeTarget == null)
+                {
+                    return;
+                }
+
+                var testMethodOrClass = c.SemanticModel.GetDeclaredSymbol(attributeTarget);
+
+                if (testMethodOrClass != null &&
+                    testMethodOrClass.GetAttributes().Any(IsTestOrTestClassAttribute) &&
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: Isn't it better to avoid querying `GetAttributes()` twice?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs (hunk 1 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: This was changed because the text was confusing for @valhristov

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs (hunk 3 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: Why did you remove that check? RSPEC still says test code.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs]: @Evangelink I don't know, it returns ImmutableArray<T> which is value type and I cannot say if it is the same instance; I suppose it is returning a new instance because `GetAttributes()` is a method...

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs [hunk 6]
```diff
@@ -1,142 +1,174 @@
...
+        [TestMethod]
+        [Ignore]
+        [WorkItem(1234)]
+        public void Foo5()
+        {
+        }
+    }
+
+    [Ignore, TestClass]
+//   ^^^^^^
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: Should this line have _Noncompliant_ on it/

Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: Good catch

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs [hunk 7]
```diff
@@ -1,142 +1,174 @@
...
+    class MsTestClass4
+    {
+    }
+}
+
+namespace Tests.Diagnostics2
+{
+    using NUnit.Framework;
+
+    class NUnitClass
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: Is there no way to end-up in the same situation with NUnit? The ignore attribute at  the class level.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs (hunk 6 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: If you have `// ^^^` the `// Noncompliant` comment is not required; you could also include the `{{Message}}` check.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs (hunk 7 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: NUnit's Ignore attribute requires message argument, same as XUnit

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs (hunk 6 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: @duncanp-sonar @Evangelink let me know if you think I should add the Noncompliant comment

Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldNotBeIgnored.cs]: Thanks. I didn't know whether it was required or not.
I think the issues are easier to see if the Noncompliant comment is there, and it seems to be normal practice to include it.
